src/jenkins.py

```diff
@@ -347,4 +375,129 @@ def add_agent_node(
+
+    items = xml_tree.findall("./channel/item")
+    # mypy doesn't understand that None type is not possible.
+    titles = (
+        item.find("title").text  # type: ignore
+        for item in items
+        if item.find("title") is not None and item.find("title").text is not None  # type: ignore
+    )
+    versions = (title.removeprefix("Jenkins ") for title in titles)  # type: ignore
+    return versions
+
+
+def get_latest_patch_version(current_version: str) -> str:
+    """Get the latest lts patch version matching with the current version.
+
+    Args:
+        current_version: Current LTS semantic version.
+
+    Returns:
+        The latest patched version available.
+
+    Raises:
+        JenkinsNetworkError: if there was an error fetching the LTS RSS feed.
+        ValidationError: if the RSS feed contains no matching LTS version.
+    """
+    try:
+        versions = _fetch_versions_from_rss()
+    except JenkinsNetworkError as exc:
+        raise JenkinsNetworkError("Failed to fetch LTS versions from RSS feed.") from exc
+
+    maj_min_version = _get_major_minor_version(current_version)
+    for version in versions:
+        if version.startswith(maj_min_version):
+            return version
+
+    raise ValidationError(
+        f"No matching version with {current_version} found from stable RSS feed."
+    )
+
+
+def download_stable_war(connectable_container: ops.Container, version: str) -> None:
+    """Download and replace the war executable.
+
+    Args:
+        connectable_container: The Jenkins container with jenkins.war executable.
+        version: Desired version of the war to download.
+
+    Raises:
+        JenkinsNetworkError: if there was an error fetching the jenkins.war executable.
+    """
+    try:
+        res = requests.get(f"{WAR_DOWNLOAD_URL}/{version}/jenkins.war", timeout=300)
+        res.raise_for_status()
+    except (
+        requests.exceptions.ConnectionError,
+        requests.exceptions.Timeout,
+        requests.exceptions.HTTPError,
+    ) as exc:
+        logger.error("Failed to download Jenkins war executable, %s", exc)
+        raise JenkinsNetworkError(f"Failed to download Jenkins war version {version}") from exc
+    connectable_container.push(
+        EXECUTABLES_PATH / "jenkins.war", res.content, encoding="utf-8", user=USER, group=GROUP
+    )
+
+
+def safe_restart(
+    credentials: Credentials, client: jenkinsapi.jenkins.Jenkins | None = None
+) -> None:
+    """Safely restart Jenkins server after all jobs are done executing.
+
+    Args:
+        credentials: The credentials of a Jenkins user with access to the Jenkins API.
+        client: The API client used to communicate with the Jenkins server.
+
+    Raises:
+        JenkinsError: if there was an API error calling safe restart.
+    """
+    client = client if client is not None else _get_client(credentials)
+    try:
+        client.safe_restart(wait_for_reboot=True)
```

Comment:
Does this method has a timeout ?
I don't think that's the case but any risk of parallel execution ? (shouldn't be the case since juju should execute 1 update-status event at a time)

Reply:
Unfortunately it doesn't provide a timeout parameter, it's a polling method wrapped without a timeout. However, the polling is:
- every 5 seconds
- max 30 repeat
I think this is a sane configuration value that can be used :)
On the risk of parallel execution, there cannot be a way unless a user calls `/safeRestart` endpoint manually :(.